### ISSUE TITLE
How to adjust verbosity to suppress thousands of lines of informational logs?

### ISSUE THREAD

**Issue by ClimbsRocks** [tensorflow/tensorflow#8340]
Per [#1258](https://github.com/tensorflow/tensorflow/issues/1258), here's my current import:

```python
import os
os.environ['TF_CPP_MIN_VLOG_LEVEL'] = '3'
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'
from keras.constraints import maxnorm
from keras.layers import Dense, Dropout
from keras.models import Sequential
from keras.wrappers.scikit_learn import KerasRegressor, KerasClassifier
```

However, as you can see if you look at a recent [Travis build](https://travis-ci.org/ClimbsRocks/auto_ml/jobs/210446239), there are still tens of thousands of lines of informational logs, like so:

```
Level 1:tensorflow:Registering Pack (<function _PackGrad at 0x7fc9c4ab3b70>) in gradient.
Level 1:tensorflow:Registering Unpack (<function _UnpackGrad at 0x7fc9c490d730>) in gradient.
Level 1:tensorflow:Registering Concat (<function _ConcatGrad at 0x7fc9c490d840>) in gradient.
Level 1:tensorflow:Registering ConcatV2 (<function _ConcatGradV2 at 0x7fc9c490d8c8>) in gradient.
Level 1:tensorflow:Registering ConcatOffset (None) in gradient.
Level 1:tensorflow:Registering Slice (<function _SliceGrad at 0x7fc9c490d9d8>) in gradient.

...

Level 1:tensorflow:  in  --> gradients_49/Relu_86_grad/ReluGrad:0
Level 1:tensorflow:  out --> gradients_49/add_896_grad/Reshape:0, gradients_49/add_896_grad/Reshape_1:0
Level 1:tensorflow:Gradient for 'MatMul_135'
Level 1:tensorflow:  in  --> gradients_49/add_896_grad/Reshape:0
Level 1:tensorflow:  out --> gradients_49/MatMul_135_grad/MatMul:0, gradients_49/MatMul_135_grad/MatMul_1:0
Level 1:tensorflow:Gradient for 'Relu_85'
Level 1:tensorflow:  in  --> gradients_49/MatMul_135_grad/MatMul:0
Level 1:tensorflow:  out --> gradients_49/Relu_85_grad/ReluGrad:0
Level 1:tensorflow:Gradient for 'add_895'
```

The [docs](https://www.tensorflow.org/api_docs/python/tf/logging/vlog) are not obvious for how to adjust this. 

***Using Python and Keras, how can I adjust verbosity to ignore these logs?***

### Environment info
Operating System:
Linux and Mac

Using v1.0.1 on both platforms, installed directly from binary URLs

**Comment by aselle** [comment]
Could you try adjusting the environment variables outside the program i.e. using export 
```sh
TF_CPP_MIN_VLOG_LEVEL=3.
```
It could be that os.environ() is not succesfully putenving the updates.

**Comment by ClimbsRocks** [comment]
Thanks for the idea, but unfortunately that didn't help either. 

I was able to at least reduce the amount of logging during training time using the following code snippet 

```
# Suppress some level of logs
os.environ['TF_CPP_MIN_VLOG_LEVEL'] = '3'
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'
from tensorflow import logging
logging.set_verbosity(logging.INFO)
from keras.constraints import maxnorm
# more imports and everything else follows below
```

The above reduces training logging, but I'm still facing several hundred lines of `tensorflow: Level 1: Registering` logging, which I assume comes when the package is first imported

```
tensorflow: Level 1: Registering FakeQuantWithMinMaxArgs (<function _FakeQuantWithMinMaxArgsGradient at 0x115e56320>) in gradient.
tensorflow: Level 1: Registering FakeQuantWithMinMaxVars (<function _FakeQuantWithMinMaxVarsGradient at 0x115ef9230>) in gradient.
tensorflow: Level 1: Registering FakeQuantWithMinMaxVarsPerChannel (<function _FakeQuantWithMinMaxVarsPerChannelGradient at 0x115ef92a8>) in gradient.
tensorflow: Level 1: Registering MatMul,flops (<function _calc_mat_mul_flops at 0x11611fcf8>) in statistical functions.
tensorflow: Level 1: Registering cond_context ((<class 'tensorflow.core.protobuf.control_flow_pb2.CondContextDef'>, <unbound method CondContext.to_proto>, <function from_proto at 0x1161c42a8>)) in proto functions.
tensorflow: Level 1: Registering while_context ((<class 'tensorflow.core.protobuf.control_flow_pb2.WhileContextDef'>, <unbound method WhileContext.to_proto>, <function from_proto at 0x1161c4b18>)) in proto functions.
tensorflow: Level 1: Registering Pack (<function _PackGrad at 0x11622bd70>) in gradient.
tensorflow: Level 1: Registering Unpack (<function _UnpackGrad at 0x11622bde8>) in gradient.
ten
``` 

The Registering logging is particularly annoying when running the test suite, because it puts several hundred lines of logs in between the info I'm scrolling between. 

Any thoughts? Can we get any signal from the fact that `logging.set_verbosity()` reduces logs for training, and use that to understanding how we might be able to suppress the Registering logging?

**Comment by drpngx** [comment]
I think the logging is in keras. Can you check where it comes from?

**Comment by ClimbsRocks** [comment]
@drpngx Thanks for the response! 

I tried commenting out all of the Keras imports in my tests, and still got the "Registering" logs.

The log looks like it's coming from https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/framework/registry.py#L64

Let me know if there's any other info I can provide to help debug this! My current hypothesis is that all this logging happens as part of the `import tensorflow as tf`, before I'm able to set_logging. 

One possible workaround is to allow the user to import some very tiny function like `set_logging` that lets them set logging defaults before importing the rest of tensorflow.

**Comment by drpngx** [comment]
OK, can you set the `vlog` to `0`, import tensorflow, then set it back to whatever you want?

**Comment by ClimbsRocks** [comment]
@drpngx : 

Just tried: 

```
    os.environ['TF_CPP_MIN_VLOG_LEVEL'] = '0'
    os.environ['TF_CPP_MIN_LOG_LEVEL'] = '0'
    from tensorflow import logging
    logging.set_verbosity(logging.INFO)
```

And still got the same Registering messages: 

```
...
Level 1:tensorflow:Registering TopK (<function _TopKGrad at 0x10c8625f0>) in gradient.
Level 1:tensorflow:Registering CTCLoss (<function _CTCLossGrad at 0x10c8626e0>) in gradient.
Level 1:tensorflow:Registering CTCGreedyDecoder (None) in gradient.
Level 1:tensorflow:Registering CTCBeamSearchDecoder (None) in gradient.
Level 1:tensorflow:Registering SetSize (None) in gradient.
Level 1:tensorflow:Registering DenseToDenseSetOperation (None) in gradient.
Level 1:tensorflow:Registering DenseToSparseSetOperation (None) in gradient.
Level 1:tensorflow:Registering SparseToSparseSetOperation (None) in gradient.
Level 1:tensorflow:Registering SdcaFprint (None) in gradient.
Level 1:tensorflow:Registering SdcaOptimizer (None) in gradient.
Level 1:tensorflow:Registering SdcaShrinkL1 (None) in gradient.
Level 1:tensorflow:Registering RandomCrop (None) in gradient.
Level 1:tensorflow:Registering RGBToHSV (None) in gradient.
Level 1:tensorflow:Registering HSVToRGB (None) in gradient.
Level 1:tensorflow:Registering DrawBoundingBoxes (None) in gradient.
Level 1:tensorflow:Registering SampleDistortedBoundingBox (None) in gradient.
Level 1:tensorflow:Registering ExtractGlimpse (None) in gradient.
Level 1:tensorflow:Registering NonMaxSuppression (None) in gradient.
Level 1:tensorflow:Registering savers ((<class 'tensorflow.core.protobuf.saver_pb2.SaverDef'>, <unbound method Saver.to_proto>, <function from_proto at 0x10c9a27d0>)) in proto functions.
Level 1:tensorflow:Registering Fact (<function call_without_requiring at 0x10bce4a28>) in default shape functions.
Level 1:tensorflow:Registering TensorSummary (None) in gradient.
Level 1:tensorflow:Registering queue_runners ((<class 'tensorflow.core.protobuf.queue_runner_pb2.QueueRunnerDef'>, <unbound method QueueRunner.to_proto>, <function from_proto at 0x10ca1cc08>)) in proto functions.
```

Is there a different way I should be setting the vlog?'

If it helps any, this is while running `nosetests -v tests` (which is what's run on the travis-build I reference above).

Thanks again for the help!

**Comment by drpngx** [comment]
OK, here's the thing:

1. The logging default is set during `tf.logging` [initialization](https://www.github.com/tensorflow/tensorflow/blob/master/tensorflow/python/platform/default/_tf_logging.py#L55)
2. You can't set the logging without triggering `tensorflow/__init__.py`, which will trigger registering all gradients.

I don't see any way to change this short of changing the code. One last ditch effort would be to try and set `sys.flags.interactive = False` but it's probably already there.

Sorry, I'm going to close this issue because I don't think it's feasible for us to fix this reasonably well. Feel free to open a new feature request or re-open if you feel strongly about that.

**Comment by CNugteren** [comment]
I know it's a bit late, but it seems this issue was never solved. I encountered the same issue when running `nosetests` and found a solution on [StackOverflow](https://stackoverflow.com/questions/43337601/nosetests-with-tensorflow-lots-of-debugging-output-how-to-disable):
```
import logging
logging.getLogger('tensorflow').disabled = True
```
Adding this before importing TensorFlow (e.g. somewhere in `__init__.py`) solves it for me.

**Comment by aselle** [comment]
Thanks @CNugteren.

**Comment by jilljenn** [comment]
Thank you @CNugteren. It's never too late.

Somehow, now I need to do: `./manage.py test <app_name>.tests`.
Not just `./manage.py test` or `./manage.py test <app_name>`, otherwise the logs will appear again.
(your snippet was included in `<app_name>/tests/__init__.py`.

**Comment by Siddhant1410** [comment]
> I know it's a bit late, but it seems this issue was never solved. I encountered the same issue when running `nosetests` and found a solution on [StackOverflow](https://stackoverflow.com/questions/43337601/nosetests-with-tensorflow-lots-of-debugging-output-how-to-disable):
> 
> ```
> import logging
> logging.getLogger('tensorflow').disabled = True
> ```
> 
> Adding this before importing TensorFlow (e.g. somewhere in `__init__.py`) solves it for me.

This did solve the Issue I've been looking for it everywhere. thank you @CNugteren

**Comment by rajcivils** [comment]
Hello I am on tensorflow2.0 and the solution given by @CNugteren  doesen't work on my pc. This is what I did - 
`import logging
logging.getLogger('tensorflow').disabled = True
import tensorflow as tf
`
After that too it is giving me a warning and it is pretty annoying.

**Comment by CodeJjang** [comment]
For whoever gets here today, setting the tensorflow logger level to warning does the job as well:

```
Import logging
logging.getLogger('tensorflow').setLevel(logging.WARNING)
```

**Comment by hoangmt** [comment]
> For whoever gets here today, setting the tensorflow logger level to warning does the job as well:
> 
> ```
> Import logging
> logging.getLogger('tensorflow').setLevel(logging.WARNING)
> ```

This works for me. Maybe it can also be applied to others, for example:

```python
import logging
logging.getLogger('absl').setLevel('ERROR')
```
